### PR TITLE
AVM: Enable pooling of logicsig execution across a group

### PR DESCRIPTION
**Pull request by jannotti**
A total of 20k opcode ticks per transaction will be available across all logicsigs in a group.

Something to spend reviewing cycles on - we now use a single EvalParams to execute all LogicSIgs in a group. Previously, we kept creating new ones.  This pretty much enshrines serial execution of LogicSigs (within group) but I don't mind that because: 1) Individual groups still seem small enough (and rarely have multiple lsigs anyway) so I don't think much optimization opportunity is lost, and 2) We already had several things that pushed in that direction - shared pointers, Tracer execution - so I don't think we'd ever try to squeeze out that parallelism anyway.

## Test Plan

One of the tricky things about this change is that MinAvmVersion is now a uint64 rather than *uint64, and it is now required to be set before calling `eval`. In order to shake out bugs in which callers did not set it, the PR was first tested by removing the check that sets the the min version inside eval, while still using a pointer, thus nil pointer errors would be triggered.

Only after all tests passed was it changed to `uint64`.

Further, EvalParams now have a `runMode` and can only be used to run apps OR sigs, not both. Since callers can't set the unexported value, they must use the exported constructors, ensure minAvmVersion and similar things are initialized.

### CODE REVIEW

File: data/transactions/verify/txn.go [hunk 1]
```diff
@@ -59,24 +59,29 @@ const txnPerWorksetThreshold = 32
...
-// For efficient caching, these parameters should either be constant
-// or change slowly over time.
-//
-// Group data are omitted because they are committed to in the
-// transaction and its ID.
+// The second set are derived from the first set and from the transaction
+// data. They are stored here only for efficiency, not for correctness, so they
+// are not checked in Equal()
 type GroupContext struct {
+	// These fields determine whether a logicsig must be re-evaluated.
```
Comment (jannotti) [inline data/transactions/verify/txn.go:73]: Notice that I've removed minAvmVersion.  Please convince yourself that was correct. My claim is that minAvmVersion can only change for a group is the consensusVersion changes (we decide to change the rules). So there's no need to explicitly track minAvmVersion too.

File: data/transactions/logic/eval.go [hunk 2]
```diff
@@ -355,58 +356,75 @@ func copyWithClearAD(txgroup []transactions.SignedTxnWithAD) []transactions.Sign
...
+		TxnGroup:             withADs,
+		Proto:                proto,
+		minAvmVersion:        computeMinAvmVersion(withADs),
+		SigLedger:            ls,
+		PooledLogicSigBudget: pooledLogicBudget,
+	}
+}
+
+// NewAppEvalParams creates an EvalParams to use while evaluating a top-level txgroup.
+func NewAppEvalParams(txgroup []transactions.SignedTxnWithAD, proto *config.ConsensusParams, specials *transactions.SpecialAddresses) *EvalParams {
```
Comment (bbroder-algo) [inline data/transactions/logic/eval.go:388]: now there's the same parameter name in two next-door functions that have different types.

File: cmd/tealdbg/local.go [hunk 3]
```diff
@@ -531,23 +531,17 @@ func (r *LocalRunner) RunAll() error {
...
-	ep := logic.NewEvalParams(txngroup, &r.proto, &transactions.SpecialAddresses{})
-	ep.SigLedger = logic.NoHeaderLedger{}
-	configureDebugger(ep)
+	sep := logic.NewSigEvalParams(r.txnGroup, &r.proto, &logic.NoHeaderLedger{})
+	aep := logic.NewAppEvalParams(txngroup, &r.proto, &transactions.SpecialAddresses{})
+	if r.debugger != nil {
+		t := logic.MakeEvalTracerDebuggerAdaptor(r.debugger)
+		sep.Tracer = t
+		aep.Tracer = t
+	}
```
Comment (bbroder-algo) [inline cmd/tealdbg/local.go:544]: better

File: data/transactions/verify/txn.go (hunk 1 above)
Comment (bbroder-algo) [reply data/transactions/verify/txn.go:73]: It seems like the min avm version only changes if there's a rekey or an appcall in the group.  Any other requirement on the minimum version would have to come from a consensus change.

File: data/transactions/logic/eval.go [hunk 4]
```diff
@@ -425,6 +443,14 @@ func (ep *EvalParams) computeAvailability() *resources {
 	return available
 }
 
+func (cx *EvalContext) minAvmVersionCheck() error {
+	if cx.version < cx.EvalParams.minAvmVersion {
+		return fmt.Errorf("program version must be >= %d for this transaction group, but have version %d",
+			cx.minAvmVersion, cx.version)
+	}
+	return nil
+}
```
Comment (bbroder-algo) [inline data/transactions/logic/eval.go:452]: broken out from `versionCheck`

File: data/transactions/logic/eval.go [hunk 5]
```diff
@@ -1350,23 +1397,21 @@ func (cx *EvalContext) step() error {
...
 
-	if cx.remainingBudget() < 0 {
-		// We're not going to execute the instruction, so give the cost back.
-		// This only matters if this is an inner ClearState - the caller should
-		// not be over debited. (Normally, failure causes total txtree failure.)
-		cx.cost -= opcost
-		if cx.PooledApplicationBudget != nil {
-			*cx.PooledApplicationBudget += opcost
-		}
+	if opcost > cx.remainingBudget() {
```
Comment (bbroder-algo) [inline data/transactions/logic/eval.go:1401]: this seems more restrictive than before?

File: data/transactions/logic/eval.go [hunk 6]
```diff
@@ -1350,23 +1397,21 @@ func (cx *EvalContext) step() error {
...
 		return fmt.Errorf("pc=%3d dynamic cost budget exceeded, executing %s: local program cost was %d",
 			cx.pc, spec.Name, cx.cost)
 	}
 
+	cx.cost += opcost
+	// Only one of these pooled budgets will be non-nil, perhaps we could collapse to one variable.
```
Comment (bbroder-algo) [inline data/transactions/logic/eval.go:1407]: can't you check cx.ep.runmode?

File: data/transactions/verify/txn.go [hunk 7]
```diff
@@ -59,24 +59,29 @@ const txnPerWorksetThreshold = 32
...
 	specAddrs        transactions.SpecialAddresses
 	consensusVersion protocol.ConsensusVersion
-	consensusParams  config.ConsensusParams
-	minAvmVersion    uint64
-	signedGroupTxns  []transactions.SignedTxn
-	ledger           logic.LedgerForSignature
```
Comment (bbroder-algo) [inline data/transactions/verify/txn.go:76]: pushed into evalparams

**Comment by bbroder-algo** [comment]
Went through the main logic, didn't see any red flags, didn't look at tests yet.

File: cmd/tealdbg/local.go (hunk 3 above)
Comment (jannotti) [reply cmd/tealdbg/local.go:544]: Yeah, the old concern about Go's weird nil handling of interfaces is gone, since we have to build the adaptor if non-nil.

File: data/transactions/logic/eval.go (hunk 5 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:1401]: It's not intended to be. Let me make the case.

Previously, we calculated the cost of the next opcode `opcost` and then performed the cost increment / budget decrement. Then we asked, "would that be going too far?" (`cx.remainingBudget() < 0`). If so, we undid the cost increment / budget decrement.

Now, we calculate that next cost `opcost` and check beforehand `opcost > cx.remainingBudget()` and fail if it is. This way, there's no need to back out a change that was illegal in the first place.

But the intent is to fail under the exact same conditions.

File: data/transactions/logic/eval.go (hunk 6 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:1407]: We could, do you like that better?  That would be a switch, and then we'd have each of these tests in the respective arm of the switch anyway. (we need to test non-nil no matter what, since pre-pooling, these budgets are also nil).

So by wrapping everything in switch arms, we strictly add code. Maybe it's clearer though?

File: data/transactions/logic/eval.go (hunk 5 above)
Comment (bbroder-algo) [reply data/transactions/logic/eval.go:1401]: i think I missed the relocation of `cx.cost +=`

File: test/scripts/e2e_subs/lsig-budget.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+# 292 is too much
+tx0 = goal.pay(escrow, escrow, amt=0, note=b'200', send=False)
+tx1 = goal.pay(escrow, escrow, amt=0, note=b'92', send=False)
+stx0 = goal.sign_with_program(tx0, code, [(200).to_bytes(8, "big")])
+stx1 = goal.sign_with_program(tx1, code, [(92).to_bytes(8, "big")])
+txinfo, err = goal.send_group([stx0, stx1])
+assert err, txinfo
+
+
+print(f"{os.path.basename(sys.argv[0])} OK {stamp}")
```
Comment (bbroder-algo) [inline test/scripts/e2e_subs/lsig-budget.py:87]: you don't want to reassign stamp to update the timestamp?

File: data/transactions/logic/eval.go (hunk 6 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:1407]: We ain't got no fancy ternary operator round these parts.

I do like avoiding the second check if the first is true.  How about:
```
switch {
	case cx.PooledApplicationBudget != nil:
		*cx.PooledApplicationBudget -= opcost
	case cx.PooledLogicSigBudget != nil {
		*cx.PooledLogicSigBudget -= opcost
}
```
(Good point to do apps first, as we care more about perf there.)

File: data/transactions/logic/eval.go (hunk 4 above)
Comment (bbroder-algo) [reply data/transactions/logic/eval.go:452]: what if you made a method func for evalcontext called setVersion you can call during cx "construction time" that sets the context version and PC given a program and proto if it conforms to the four version requirements, and get rid of versionCHeck and minAvmVersionCheck.

File: test/scripts/e2e_subs/lsig-budget.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+tx1.group = None
+txinfo, err = goal.send_group([stx1, stx0])  # rearrange
+assert not err, err
+
+# 292 is too much
+tx0 = goal.pay(escrow, escrow, amt=0, note=b'200', send=False)
+tx1 = goal.pay(escrow, escrow, amt=0, note=b'92', send=False)
+stx0 = goal.sign_with_program(tx0, code, [(200).to_bytes(8, "big")])
+stx1 = goal.sign_with_program(tx1, code, [(92).to_bytes(8, "big")])
+txinfo, err = goal.send_group([stx0, stx1])
```
Comment (bbroder-algo) [inline test/scripts/e2e_subs/lsig-budget.py:83]: order doesn't matter here either but maybe this one doesn't rise to reorder test worthy.

File: data/transactions/logic/eval_test.go [hunk 10]
```diff
@@ -1959,50 +2004,6 @@ return
 	testLogicBytes(t, ops.Program, ep)
 }
 
-func TestGaid(t *testing.T) {
-	partitiontest.PartitionTest(t)
-
```
Comment (bbroder-algo) [inline data/transactions/logic/eval_test.go:1964]: relocated to evalStateful_test

File: data/transactions/logic/eval_test.go [hunk 11]
```diff
@@ -5061,116 +5092,6 @@ func TestBytesConversions(t *testing.T) {
 	testAccepts(t, "byte 0x0011; byte 0x10; b+; btoi; int 0x21; ==", 4)
 }
 
-func TestLog(t *testing.T) {
-	partitiontest.PartitionTest(t)
```
Comment (bbroder-algo) [inline data/transactions/logic/eval_test.go:5065]: relocated to evalStateful_test

File: data/transactions/logic/eval_test.go [hunk 12]
```diff
@@ -4208,26 +4256,6 @@ func TestArgType(t *testing.T) {
 	require.Equal(t, avmUint64, sv.avmType())
 }
 
-func TestApplicationsDisallowOldTeal(t *testing.T) {
-	partitiontest.PartitionTest(t)
```
Comment (bbroder-algo) [inline data/transactions/logic/eval_test.go:4212]: moved to evalStateful_test

File: data/transactions/logic/eval_test.go (hunk 11 above)
Comment (jannotti) [reply data/transactions/logic/eval_test.go:5065]: FWIW, the reason I did this churn was to discourage the use of `makeSampleEnv()` in eval_test.go. It creates an eval params for evaluating apps, which these tests do indeed need, but I noticed several tests in `eval_test.go` that _don't_ need it trying to use them and (now) failing because they were only testing logicsigs.

File: data/transactions/verify/txn.go [hunk 13]
```diff
@@ -125,40 +130,61 @@ func (e *TxGroupError) Unwrap() error {
 	return e.err
 }
 
-// PrepareGroupContext prepares a verification group parameter object for a given transaction
-// group.
-func PrepareGroupContext(group []transactions.SignedTxn, contextHdr *bookkeeping.BlockHeader, ledger logic.LedgerForSignature) (*GroupContext, error) {
+// verifyCount gives the number of signature verifications
+func verifyCount(txn *transactions.SignedTxn) int {
```
[the remark this answers is not in the document]
Comment (bbroder-algo) [reply data/transactions/verify/txn.go:134]: could be handy to have around.  but then maybe it should have a test.

File: data/transactions/logic/eval.go [hunk 14]
```diff
@@ -355,58 +356,75 @@ func copyWithClearAD(txgroup []transactions.SignedTxnWithAD) []transactions.Sign
...
 
-	// Make a simpler EvalParams that is good enough to evaluate LogicSigs.
-	if apps == 0 {
-		return &EvalParams{
-			TxnGroup: txgroup,
-			Proto:    proto,
-			Specials: specials,
+	var pooledLogicBudget *int
+	if lsigs > 0 { // don't allocate if no lsigs
+		if proto.EnableLogicSigCostPooling {
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/eval.go:370]: Yes, that is my intent. Willing to hear objections, of course.

File: data/transactions/logic/eval.go [hunk 15]
```diff
@@ -355,58 +356,75 @@ func copyWithClearAD(txgroup []transactions.SignedTxnWithAD) []transactions.Sign
...
+			pooledLogicBudget = new(int)
+			*pooledLogicBudget = len(txgroup) * int(proto.LogicSigMaxCost)
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/eval.go:372]: My current thinking is that it created more complexity than it's worth, given that the base problem still exists.

File: data/transactions/verify/txn.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/verify/txn.go:134]: I'll remove it. I think I left it there while I was still considering the verify penalty.

File: data/transactions/logic/evalStateful_test.go [hunk 16]
```diff
@@ -461,13 +479,20 @@ func testAppsBytes(t *testing.T, programs [][]byte, ep *EvalParams, expected ...
 
 func testApp(t *testing.T, program string, ep *EvalParams, problems ...string) transactions.EvalDelta {
 	t.Helper()
+	if ep == nil {
+		ep = defaultAppParamsWithVersion(LogicVersion)
+	}
```
Comment (jasonpaulos) [inline data/transactions/logic/evalStateful_test.go:484]: Should this call `ep.reset()` if it's not nil? This function appears to be called multiple times with the same `ep` on different programs.

Comment (jasonpaulos) [reply data/transactions/logic/evalStateful_test.go:484]: nvm, this just calls into `testAppBytes`, which does reset the `ep`

File: data/transactions/logic/export_test.go [hunk 17]
```diff
@@ -40,23 +41,29 @@ func (l *Ledger) DelBoxes(app basics.AppIndex, names ...string) {
...
 var TestAppBytes = testAppBytes
-var TestApps = testApps
 var TestLogicRange = testLogicRange
 var TestProg = testProg
 var WithPanicOpcode = withPanicOpcode
 
+// TryApps exports "testApps" while accepting a simple uint64. Annoying, we
+// can't export this as "TestApps" because it looks like a Test function with
+// the wrong signature.
+func TryApps(t *testing.T, programs []string, txgroup []transactions.SignedTxn, ver uint64, ledger *Ledger, expected ...expect) *EvalParams {
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/export_test.go:63]: Oh, ha! I bet that would work.  I think I'll do that, to shrink the overall size of PR even though it's more churn for current reviewers.

**Review by jasonpaulos — DISMISSED** [review]
A few minor comments here and there, but looks good to me. Good testing as well.

File: data/transactions/logic/eval_test.go [hunk 18]
```diff
@@ -1880,25 +1923,27 @@ func TestTxn(t *testing.T) {
...
 				// Starting in txnEffectsVersion, there are fields we can't access in Logic mode
-				testLogicFull(t, ops.Program, 3, ep, "not allowed in current mode")
+				testLogicFull(t, ops.Program, 3, sep, "not allowed in current mode")
 				// And the early tests use "arg" a lot - not allowed in stateful. So remove those tests.
 				lastArg := strings.Index(source, "arg 10\n==\n&&")
 				require.NotEqual(t, -1, lastArg)
 
-				appSafe := "int 1" + strings.Replace(source[lastArg+12:], `txn Sender
+				if v >= appsEnabledVersion {
+					aep.TxnGroup[2].EvalDelta.Logs = []string{"x", "prefilled"}
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/eval_test.go:1937]: The guard isn't needed anymore. I originally added it because of the complicated way I was constructing a single `ep` that was supposed to be good for apps and sigs.  I added an app call in, so I would get a full `ep`, but only if the version >= appsEnabled (so I wouldn't create _other_ problems). But now that's simplified, so even though apps aren't enabled in a version, I create a usable `ep`.

In a way, it's nonsense - why test how apps work in a version before they are runnable?  But the extra check just complicates the flow, so I'll remove it.

[the remark this answers is not in the document]
Comment (jasonpaulos) [reply data/transactions/logic/eval_test.go:1937]: My question was more about the `aep.TxnGroup[2].EvalDelta.Logs = []string{"x", "prefilled"}` line

[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/eval_test.go:1937]: Oh, I'll look more closely, but I assume that was so `gitxn Logs` could be tested.

But I'm finding that entire Replace suspicious. Digging in a bit.

[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/eval_test.go:1937]: It's so `gtxn 2 NumLogs` works.
(not `gitxn` as there is no inner here)

And the Replace is fine. non sequitor.

[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/eval_test.go:1937]: Added a small comment to explain.

File: test/scripts/e2e_subs/lsig-budget.py (hunk 9 above)
Comment (jannotti) [reply test/scripts/e2e_subs/lsig-budget.py:83]: yes, I consider the one test for re-ordering, and the other for overflow

File: data/transactions/logic/eval.go (hunk 2 above)
Comment (jannotti) [reply data/transactions/logic/eval.go:388]: It seemed weird to demand that lsig ep construction callers wrap the SignedTxns into SignedTxnWIthADs before calling, as ADs are about the stateful part of evaluation.  Those callers don't know about ADs.

On the other hand, for app evaluation, the callers have SignedTxnsWithAD, and I didn't want to make them copy them into a plain `[]SignedTxn`, since I would then have to re-copy anyway, since the evaluator needs the AD space to work with.

File: data/transactions/verify/txn.go (hunk 1 above)
Comment (jannotti) [reply data/transactions/verify/txn.go:73]: So basically put all of this:
```
	version, pc, err := versionCheck(program, params.Proto)
	if err != nil {
		return err
	}

	var cx EvalContext
	cx.version = version
	cx.pc = pc
...
...
	if err := cx.minAvmVersionCheck(); err != nil {
		return err
	}
```
into a cx method?

I like that, assuming it also works for `check()`, not just `eval()`. Let me look into it.

File: data/transactions/logic/eval.go (hunk 4 above)
Comment (jannotti) [reply data/transactions/logic/eval.go]: So basically put all of this:
```
	version, vlen, err := versionCheck(program, cx.Proto)

	cx.version = vlen
	cx.pc = pc
...
...
	if err := cx.minAvmVersionCheck(); err != nil {
		return false, err
	}
```
into a cx method?

See if you like https://github.com/algorand/go-algorand/pull/5528/commits/ed59b5daef014f8ce72df9e0f5d7b2b5dbe9fda9

**Review by jasonpaulos — APPROVED** [review]

**Review by bbroder-algo — APPROVED** [review]
